src/libs/Network/NetworkStore.js

```diff
@@ -0,0 +1,85 @@
+import lodashGet from 'lodash/get';
+import Onyx from 'react-native-onyx';
+import _ from 'underscore';
+import ONYXKEYS from '../../ONYXKEYS';
+
+let credentials;
+let authToken;
+let currentUserEmail;
+let networkReady = false;
+
+/**
+ * @param {Boolean} ready
+ */
+function setIsReady(ready) {
+    networkReady = ready;
+}
+
+function checkRequiredDataAndSetNetworkReady() {
+    if (_.isUndefined(authToken) || _.isUndefined(credentials)) {
+        return;
+    }
+
+    setIsReady(true);
+}
+
+Onyx.connect({
+    key: ONYXKEYS.SESSION,
+    callback: (val) => {
+        authToken = lodashGet(val, 'authToken', null);
+        currentUserEmail = lodashGet(val, 'email', null);
+        checkRequiredDataAndSetNetworkReady();
+    },
+});
+
+Onyx.connect({
+    key: ONYXKEYS.CREDENTIALS,
+    callback: (val) => {
+        credentials = val || null;
+        checkRequiredDataAndSetNetworkReady();
+    },
+});
+
+/**
+ * @returns {String}
+ */
+function getAuthToken() {
+    return authToken;
+}
+
+/**
+ * @param {String} newAuthToken
+ */
+function setAuthToken(newAuthToken) {
```

Comment:
Good eye. It's not quite a hack, but close. Setting values in Onyx always takes at least one tick of the event loop and updates subscribers late. If you look at where this is used it might make more sense.

https://github.com/Expensify/App/blob/9c6c2a9441c1f8b51bccd8a7ee4a79058f5b5d95/src/libs/API.js#L236-L237

Inside `reauthenticate()` we are setting the "local" authToken since anything called after `reauthenticate()` will need _immediate access_ to the `authToken` - otherwise it will use the old one again and get a `407` again.

I'll leave a comment here. Basically, Onyx being Onyx though.

Reply:
Thanks, makes sense.

> Setting values in Onyx always takes at least one tick of the event loop and updates subscribers late ... Basically, Onyx being Onyx though
Is there a way we could fix this? Maybe we could create an issue in the Onyx repo?

Reply:
Yeah, I think we even tried once, but it broke things in unexpected ways and there is a high likelihood that it may break stuff again. Anyone using `Onyx.merge()` is expecting to have those values available on the next tick. Our current tests are pretty dependent on that "feature" as well so it's gonna be interesting if we ever try and do anything about it.

Reply:
Hmmm seems to me like things might be a bit easier to reason about if one could safely assume than any value written in Onyx will be updated immediately, but I'll take your word for it.

Reply:
They would 1000% be easier. And we could still do it. But whoever does it will need to check everything carefully and probably remove a lot of `waitForPromisesToResolve()` calls from our tests 😂

Reply:
Created an issue: https://github.com/Expensify/react-native-onyx/issues/123

Will explore more when I have some spare cycles.